src/lib/support/Span.h

```diff
@@ -38,6 +39,12 @@ class Span
 
     const T * data() const { return mDataBuf; }
     size_t size() const { return mDataLen; }
+    bool empty() const { return size() == 0; }
+    bool data_equal(const Span & other) const
+    {
+        return (empty() && other.empty()) ||
+            (!empty() && !other.empty() && (size() == other.size()) && (memcmp(data(), other.data(), size() * sizeof(T)) == 0));
```

Comment:
So the thing is, `size() == other.size()` implies `empty() == other.empty()`, hence my suggestion to just check for size equality, then if that succeeds check for `(empty() || memcmp())`.  It ends up with simpler logic, imo.  I can live with the logic here, though I would be curious what compilers do with these in practice and which one ends up with more code.

Reply:
Nice brainteasers! updated